Refactor fee estimation

ManfredKarrer opened the pull request:
Danger zone ;-)
This needs some good testing and careful review.

ManfredKarrer committed:
Add TxFeeEstimationService

ManfredKarrer committed:
Add TxFeeEstimationServiceTest

ManfredKarrer committed:
Add getEstimatedFeeTxSize method

ManfredKarrer committed:
Remove estimateBtcTradingFeeTxSize

ManfredKarrer committed:
Use TxFeeEstimationService

- feeTxSizeEstimationRecursionCounter was never reset, so that caused
 probably bugs by ignoring fee estimation after it was called >10 times.
 The default size was used then so the bug was not very obvious as long
 the tx size was not very large.

ManfredKarrer committed:
Fix missing param in test class

blabno commented on core/src/main/java/bisq/core/btc/TxFeeEstimationService.java:
```diff
@@ -0,0 +1,159 @@
+
+        if (!preferences.isPayFeeInBtc()) {
+            // If we pay the fee in BSQ we have one input more which adds about 150 bytes
+            estimatedTxSize += 150;
```
Can we avoid magic numbers? https://en.wikipedia.org/wiki/Magic_number_(programming)

ManfredKarrer replied:
Yes we can use a constant. It is just used once so thats why I did not use one and the class is small so easty to overlook all as well there is a comment decribing the value. But no problem to make a constant for it.

blabno commented on core/src/test/java/bisq/core/btc/TxFeeEstimationServiceTest.java:
```diff
@@ -0,0 +1,123 @@
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+@RunWith(PowerMockRunner.class)
```
There is no need for PowerMock in this test class.

ManfredKarrer replied:
OK, I will remove it

blabno commented on core/src/test/java/bisq/core/btc/TxFeeEstimationServiceTest.java:
```diff
@@ -0,0 +1,123 @@
+        assertEquals(260, result);
+
+
+        // TODO check how to use the mocking framework for repeated calls
```
This is separate test scenario and should be placed in separate method.
Good naming convention for test methods is `testMethodName_scenario_expectedCondition`
i.e. `getEstimatedTxSize_nullOutputs_throwsException`

ManfredKarrer replied:
You refer to the commented out lines, right? Would you mind to have a hands on the test class? More your domain, you know my love for tests is limited ;-)

ManfredKarrer replied:
@blabno I broke them up in https://github.com/bisq-network/bisq/pull/2251/commits/877d0ae3acb2a1ea94e389f99cbd0194ebae6aee and looked again into it. See commit comments... Could u have a look to try to fix it?

blabno commented on core/src/main/java/bisq/core/btc/TxFeeEstimationService.java:
```diff
@@ -0,0 +1,159 @@
+        return new Tuple2<>(txFee, size);
+    }
+
+    @VisibleForTesting
```
I've been always told that this is a bad idea, as it's an attempt to test implementation details.
Either extract it to an utility class that could be tested separately or do not test it individually.

ManfredKarrer replied:
Not clear what you mean. Lets discuss on a call in a bit...

ripcurlx replied:
I think @blabno is talking about the fact, that if you need to make something visible for testing that wouldn't be accessible otherwise, is a sign that something should be structured or tested differently 😉

ManfredKarrer replied:
Ah ok. Yes basically agree. In that case I would not like to expose getEstimatedTxSize as it is only interesting for that class. The more high level methods carry much more special case handling that they would be harder to test then the core functionality in getEstimatedTxSize.

ManfredKarrer committed:
Remove PowerMockRunner

ManfredKarrer committed:
Use constants for tx size values

ManfredKarrer committed:
Break up test method

Set ignore to failing tests

@Bernard could you have a look and try to get those working?
Reason for problem is that repeated calls to getEstimatedFeeTxSize
do not work (returns 0 at second call in loop which cause test to fail)

ripcurlx commented on core/src/main/java/bisq/core/btc/TxFeeEstimationService.java:
```diff
@@ -0,0 +1,163 @@
+    private static int DEPOSIT_TX_SIZE = 320;
+    private static int PAYOUT_TX_SIZE = 380;
+    private static int BSQ_INPUT_INCREASE = 150;
+    private static int MAX_ITERATIONS = 10;
```
Is there a specific reason why you've put payout_tx_size, bsq_input_increase and max_iterations here and not as a local variable? To get a better overview on the values used within this class?

ManfredKarrer replied:
I have not a very strong opinion here, @blabno critiziced that they should be statics and not magic numbers in the code. I usually only make statics if they are used repeatedly and/or if the class is longer and it helps to see quickly important parameters. I would prefer to leave it now.

ManfredKarrer committed:
Add comment to give more background about the fee estimation

blabno commented on core/src/test/java/bisq/core/btc/TxFeeEstimationServiceTest.java:
```diff
@@ -0,0 +1,146 @@
+        realTxSize = 2600;
+
+        txFee = txFeePerByte.multiply(initialEstimatedTxSize);
+        when(btcWalletService.getEstimatedFeeTxSize(outputValues, txFee)).thenReturn(realTxSize);
```
do/while loop inside `getEstimatedFeeTxSize` has 2 cycles because first invocation of `isInTolerance` returns `false`.
So for the first iteration we have to mock with initial `txFee`:
        
    when(btcWalletService.getEstimatedFeeTxSize(outputValues, txFee)).thenReturn(realTxSize);

And for the second iteration we have to mock with `txFee` multiplied by `txFeePerByte`.

    when(btcWalletService.getEstimatedFeeTxSize(outputValues, txFeePerByte.multiply(realTxSize))).thenReturn(realTxSize)

Alternatively we can mock `getEstimatedFeeTxSize` to return the same result for any `txFee`:

    when(btcWalletService.getEstimatedFeeTxSize(eq(outputValues), any(Coin.class))).thenReturn(realTxSize);

blabno commented on core/src/test/java/bisq/core/btc/TxFeeEstimationServiceTest.java:
```diff
@@ -0,0 +1,146 @@
+        assertEquals(2600, result);
+    }
+
+    // FIXME @Bernard could you have a look?
```
Same as above

ManfredKarrer replied:
@blabno Thanks. I would prefer to leave that for another PR, too buys with other high prio stuff. If you have time to take that would be great as well!

ManfredKarrer commented:
@ripcurlx Could you ACK if you have completed review?

ripcurlx commented on core/src/main/java/bisq/core/btc/TxFeeEstimationService.java:
```diff
@@ -0,0 +1,179 @@
+        int estimatedTxSize = initialEstimatedTxSize;
+        int realTxSize;
+        int counter = 0;
+        do {
```
As discussed it would be good to add a note why there is the need to run this part of the code up to 10 times to find the best transaction setup.

ManfredKarrer replied:
OK, I will add more comment.

ripcurlx commented on desktop/src/main/java/bisq/desktop/main/offer/MutableOfferDataModel.java:
```diff
@@ -135,10 +134,9 @@
-    private int feeTxSizeEstimationRecursionCounter;
+    protected Coin txFeeFromFeeService = Coin.ZERO;
+    protected boolean marketPriceAvailable;
+    protected int feeTxSize = TxFeeEstimationService.TYPICAL_TX_WITH_1_INPUT_SIZE;
```
Is there a reason why these three fields are protected and not private?

ManfredKarrer replied:
No, I will make them private.

ripcurlx commented on desktop/src/main/java/bisq/desktop/main/offer/takeoffer/TakeOfferDataModel.java:
```diff
@@ -328,81 +333,37 @@ void onTakeOffer(TradeResultHandler tradeResultHandler) {
         } else {
-            feeTxSize = 320;
-            txFeeFromFeeService = getTxFeeBySize(feeTxSize);
+            feeTxSize = 380;
```
Why did this change from 320?

ManfredKarrer replied:
In the TxFeeEstimationService we use as min. the payout tx size (380), so that reflects the same logic. It is to be more on the safe side and prefer to overpay the other txs and not underpay the payout tx.

From: TxFeeEstimationService
   int averageSize = (estimatedTxSize + DEPOSIT_TX_SIZE) / 2;  // deposit tx has about 320 bytes
            // We use at least the size of the payout tx to not underpay at payout.
            size = Math.max(PAYOUT_TX_SIZE, averageSize);

ripcurlx commented on desktop/src/main/java/bisq/desktop/main/offer/takeoffer/TakeOfferDataModel.java:
```diff
@@ -620,10 +581,32 @@ public String getCurrencyNameAndCode() {
     }
 
     public Coin getTotalTxFee() {
+        Coin totalTxTees = txFeeFromFeeService.add(getTxFeeForDepositTx()).add(getTxFeeForPayoutTx());
```
Typo: Should be totalTxFees not ..Tees

ManfredKarrer replied:
Fixed

ripcurlx reviewed (approved):
ACK - Just a couple of questions and minor changes.

ManfredKarrer committed:
Improve comment

ManfredKarrer committed:
Fix typos

ManfredKarrer committed:
Make fields private

ManfredKarrer commented:
No idea why tests fail, the EditOfferDataModelTest does not exist. Seems a caching issue.
